### PR TITLE
iochaos: add selector for injecting sidecar when using init-request config

### PR DESCRIPTION
**Pull request by cwen0**
Signed-off-by: cwen0 <cwenyin0@gmail.com> 

In this pr, we add a selector in the config of the sidecar, because we need it when we inject sidecar at the starting of the app cluster.  For example, we want to inject sidecar to tikv pod at the starting of tidb cluster, we can apply the config of tikv sidecar  [ tikv-configmap.yaml](https://github.com/pingcap/chaos-mesh/blob/master/examples/chaosfs-configmap/tikv-configmap.yaml) and the selector need to add to the config file.

eg: 
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: chaosfs-tikv
  labels:
    app.kubernetes.io/component: webhook
data:
  chaosfs-tikv.yaml: |
    name: chaosfs-tikv
    selector: 
      labelSelectors:
        "app.kubernetes.io/component": "tikv"
    initContainers:
    - name: inject-scripts
      image: pingcap/chaos-scripts:latest
      imagePullpolicy: Always
      command: ["sh", "-c", "mkdir -p /tmp/scripts; cp -R /scripts/* /tmp/scripts/; mkdir -p /var/lib/tikv/fuse-data"]
    containers:
   .....
``` 

Then add `admission-webhook.pingcap.com/init-request=chaosfs-tikv` to the annotation of the tidb cluster namespace and start the tidb cluster. The tikv pod created next will be injected into the sidecar container.

### CODE REVIEW

File: pkg/utils/selector.go [hunk 1]
```diff
@@ -127,6 +127,64 @@ func SelectPods(ctx context.Context, c client.Client, selector v1alpha1.Selector
...
+// TODO: support to check fieldsSelector
+func CheckPodMeetSelector(pod v1.Pod, selector v1alpha1.SelectorSpec) (bool, error) {
+	if len(selector.Pods) > 0 {
+		for ns, names := range selector.Pods {
+			if pod.Namespace != ns {
+				continue
+			}
+
+			for _, name := range names {
+				if pod.Name == name {
```
Comment (zhouqiang-cl) [inline pkg/utils/selector.go]: Did we need to mind regex such as *

File: pkg/webhook/config/config.go [hunk 2]
```diff
@@ -24,6 +24,7 @@ import (
 	"sync"
 
 	"github.com/ghodss/yaml"
+	"github.com/pingcap/chaos-mesh/api/v1alpha1"
```
Comment (zhouqiang-cl) [inline pkg/webhook/config/config.go]: Did we can change to v1

File: pkg/utils/selector.go (hunk 1 above)
Comment (cwen0) [reply pkg/utils/selector.go]: I think `regex` is not necessary at present.

File: pkg/webhook/config/config.go (hunk 2 above)
Comment (cwen0) [reply pkg/webhook/config/config.go]: Too early 😂

File: pkg/utils/selector.go [hunk 3]
```diff
@@ -127,6 +127,64 @@ func SelectPods(ctx context.Context, c client.Client, selector v1alpha1.Selector
...
+					return true, nil
```
Comment (mahjonp) [inline pkg/utils/selector.go]: why return true once pod name matches? maybe other selectors(eg.`AnnotationSelectors`) don't match

File: pkg/webhook/inject/inject.go [hunk 4]
```diff
@@ -84,6 +84,24 @@ func Inject(res *v1beta1.AdmissionRequest, cli client.Client, cfg *config.Config
 		}
 	}
 
+	if injectionConfig.Selector != nil {
+		meet, err := utils.CheckPodMeetSelector(pod, *injectionConfig.Selector)
+		if err != nil {
+			log.Error(err, "Failed to check pod selector", "namespace", pod.Namespace)
+			return &v1beta1.AdmissionResponse{
+				Allowed: true,
```
[the remark this answers is not in the document]
Comment (cwen0) [reply pkg/webhook/inject/inject.go]: this response will ignore injection and just to create pod

File: pkg/utils/selector.go (hunk 3 above)
Comment (cwen0) [reply pkg/utils/selector.go]: yep, you are right. I will fix it.

Comment (cwen0) [reply pkg/utils/selector.go]: oh, if the pod met this selector, it means that the pod is specified.

Comment (cwen0) [reply pkg/utils/selector.go]: This may be ambiguous, let me fix it

Comment (mahjonp) [reply pkg/utils/selector.go]: we'd better let the selection continues?

Comment (cwen0) [reply pkg/utils/selector.go]: yep, you are right, I will fix it.

**Comment by cwen0** [comment]
@mahjonp PTAL again, thanks!

File: pkg/utils/selector.go [hunk 5]
```diff
@@ -131,17 +131,24 @@ func SelectPods(ctx context.Context, c client.Client, selector v1alpha1.Selector
...
 			for _, name := range names {
 				if pod.Name == name {
-					return true, nil
+					meet = true
+					break Loop
 				}
 			}
 		}
+
+		if !meet {
```
Comment (mahjonp) [inline pkg/utils/selector.go]: Don't need `Loop` label. when leaving the inner for-block, if meet is false boolean, it has no chance to become true.

**Pull request merged**